kani-compiler/src/codegen_cprover_gotoc/codegen/contract.rs

```diff
@@ -35,8 +35,45 @@ impl<'tcx> GotocCtx<'tcx> {
     ) -> AssignsContract {
         let tcx = self.tcx;
         let function_under_contract_attrs = KaniAttributes::for_item(tcx, function_under_contract);
-        let wrapped_fn = function_under_contract_attrs.inner_check().unwrap().unwrap();
 
+        let recursion_wrapper_id =
+            function_under_contract_attrs.checked_with_id().unwrap().unwrap();
+        let mut recursion_tracker = items.iter().filter_map(|i| match i {
+            MonoItem::Static(recursion_tracker)
+                if (*recursion_tracker).name().contains(
+                    format!("{}::REENTRY", tcx.item_name(recursion_wrapper_id)).as_str(),
```

Comment:
This is much better, though I wonder if instead of checking for `contains` we could check directly for equality? And also, as a small optimization, I would move the `format` outside the loop. E.g.

```rs
let expecterd_name = format!("{}::REENTRY", tcx.item_name(recursion_wrapper_id);
let mut recursion_tracker = items.iter().filter_map(|i| match i {
    MonoItem::Static(recursion_tracker)
        if (*recursion_tracker).name() == expected_name.as_str() =>
    {
        Some(*recursion_tracker)
    }
    _ => None,
});

Reply:
We can't check for directly equality because `tcx.item_name(recursion_wrapper_id)` doesn't give us the module of the function, so there is still an information missing and that's why I used contains. But I can definitely move the `format` outside of the loop.